lib/Distribution/Helper/Discover.hs

```diff
@@ -54,6 +56,14 @@ findProjects dir = execWriterT $ do
   let stackYaml = dir </> "stack.yaml"
   whenM (liftIO $ doesFileExist stackYaml) $
     tell [Ex $ ProjLocStackYaml stackYaml]
+
+  cabalVersion <- liftIO $ let ?verbose = (<=0)
+                               ?progs = defaultPrograms
```

Comment:
Now this is just a no-go and I think it breaks the patch because I don't want to add a progs argument to this function.

IMO the return value here shouldn't depend on the runtime version of cabal but we should always return all possible projects and the API clients get to filter out what they want. We could offer a convinience function to do that though, something like `_ :: SProjType pt -> [Ex ProjLoc] -> [ProjLoc pt]` whch also happens to nicely re-expose the type-level information of which project type you're dealing with.

Though come to think of it, maybe it would be better altogether to just have `findProjects :: SProjType pt -> FilePath -> [ProjLoc pt]` in the first place. 

Regardless, in that framework I don't think we actually need to know cabal's version at all because c-h only supports versions that already have v2-build support so the API client choosing to prefer a v2 project context is always going to be a valid choice.

Reply:
FYI I'm working on a refactor that will unify the `ProjLocV{1,2}{File,Dir}` ProjLocs. In the initial design I didn't realise v2-build also supports just a plain cabal file, though of course it does for backwards compatibility, so this isn't really modelled well in the API hence the need for this cludge in the discovery code.

Reply:
@DanielG so then would it make sense to just not check the version at all, and always return the v2 project?